{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1758a8ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Looking in indexes: https://pypi.org/simple, https://pypi.ngc.nvidia.com\n",
      "Requirement already satisfied: biopython in /home/blessyantony/.local/lib/python3.8/site-packages (1.83)\n",
      "Requirement already satisfied: numpy in /opt/conda/lib/python3.8/site-packages (from biopython) (1.21.4)\n"
     ]
    }
   ],
   "source": [
    "! pip install biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "af85c69f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from ast import literal_eval\n",
    "import unicodedata\n",
    "import matplotlib.pyplot as plt\n",
    "from Bio import SeqIO\n",
    "pd.set_option(\"display.max_rows\", None)\n",
    "pd.set_option('display.width', 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "de798ac2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation',\n",
       " '/opt/conda/lib/python38.zip',\n",
       " '/opt/conda/lib/python3.8',\n",
       " '/opt/conda/lib/python3.8/lib-dynload',\n",
       " '',\n",
       " '/home/blessyantony/.local/lib/python3.8/site-packages',\n",
       " '/opt/conda/lib/python3.8/site-packages',\n",
       " '/opt/conda/lib/python3.8/site-packages/IPython/extensions',\n",
       " '/home/blessyantony/.ipython',\n",
       " '/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../../../../..',\n",
       " '/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../../../..',\n",
       " '/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../../..',\n",
       " '/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../..']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "sys.path.append(os.path.join(os.getcwd(), \"..\", \"..\", \"..\", \"..\", \"..\"))\n",
    "sys.path.append(os.path.join(os.getcwd(), \"..\", \"..\", \"..\", \"..\"))\n",
    "sys.path.append(os.path.join(os.getcwd(), \"..\", \"..\", \"..\"))\n",
    "sys.path.append(os.path.join(os.getcwd(), \"..\", \"..\"))\n",
    "sys.path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "574a913d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def column_stats(df, column_name, n=None):\n",
    "    if n is None:\n",
    "        n = df.shape[0]\n",
    "    count_column_name = column_name + \"_count\"\n",
    "    count_df = pd.DataFrame(df[column_name].value_counts()).reset_index()\n",
    "    count_df.columns=[column_name, count_column_name]\n",
    "    count_df[column_name + \"_percent\"] = count_df[count_column_name].apply(lambda x: int(x)/n*100)\n",
    "    print(f\"Number of unique values = {len(df[column_name].unique())}\")\n",
    "    print(f\"{count_df}\")\n",
    "    return count_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "05e2a3dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_seq_len_histogram(df, n_bins):\n",
    "    sns.histplot(df[\"seq_len\"])\n",
    "    print(f\"min seq len = {min(df['seq_len'])}\")\n",
    "    print(f\"max seq len = {max(df['seq_len'])}\")\n",
    "    plt.show()\n",
    "    \n",
    "    freq, bins = np.histogram(df[\"seq_len\"], bins=n_bins)\n",
    "    n = df.shape[0]\n",
    "    hist_map = []\n",
    "    for i in range(n_bins):\n",
    "        hist_map.append({\"start\": bins[i], \"end\":bins[i+1], \"count\": freq[i], \"percentage\": freq[i]/n*100})\n",
    "    hist_df = pd.DataFrame(hist_map)\n",
    "    print(hist_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "5bf971ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_uniref_fasta_file(input_file_path):\n",
    "    sequences = []\n",
    "    i = 0\n",
    "    print(\"START: Parsing fasta file\")\n",
    "    print(input_file_path)\n",
    "    # parse fasta file to extract uniref90_id, tax_id of virus/organism, and protein sequence\n",
    "    with open(input_file_path) as f:\n",
    "        for record in SeqIO.parse(f, \"fasta\"):\n",
    "            i += 1\n",
    "            sequences.append({\"uniprot_id\": record.id, \"seq\": str(record.seq)})\n",
    "\n",
    "    print(\"END: Parsing fasta file\")\n",
    "    print(len(sequences))\n",
    "    print(f\"Number of records parsed = {i}\")\n",
    "    return pd.DataFrame(sequences)\n",
    "    \n",
    "\n",
    "columns = [\"uniprot_id\", \"seq\", \"virus_host_name\", \"human_binary_label\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "bc239dcc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>virus_name</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>virus_host_name</th>\n",
       "      <th>human_binary_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WIV04</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>WIV04(MN996528.1) Wuhan variant index virus</td>\n",
       "      <td>1273</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  uniprot_id                                                seq                                   virus_name  seq_len virus_host_name human_binary_label\n",
       "0      WIV04  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...  WIV04(MN996528.1) Wuhan variant index virus     1273    homo sapiens       homo sapiens"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wiv04_sequence_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/wiv04/sarscov2-S-WIV04ref.csv\")\n",
    "wiv04_df = pd.read_csv(wiv04_sequence_file_path)\n",
    "wiv04_df.rename(columns={\"uniref90_id\": \"uniprot_id\"}, inplace=True)\n",
    "wiv04_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2e962e0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3567, 12)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>uniprot_host_tax_ids</th>\n",
       "      <th>embl_ref_id</th>\n",
       "      <th>embl_host_name</th>\n",
       "      <th>virus_host_name</th>\n",
       "      <th>virus_name</th>\n",
       "      <th>virus_taxon_rank</th>\n",
       "      <th>virus_host_tax_id</th>\n",
       "      <th>virus_host_taxon_rank</th>\n",
       "      <th>seq</th>\n",
       "      <th>human_binary_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A6G8HJ43</td>\n",
       "      <td>28295</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QIM40843.1</td>\n",
       "      <td>['pig']</td>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>Porcine epidemic diarrhea virus</td>\n",
       "      <td>species</td>\n",
       "      <td>9823</td>\n",
       "      <td>species</td>\n",
       "      <td>MKSLTYFWLFLPVLSTLSLPQDVTRCSANTNFRRFFSKFNVQAPAV...</td>\n",
       "      <td>NOT homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0G3FCN8</td>\n",
       "      <td>28295</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AKJ85723.1</td>\n",
       "      <td>['swine']</td>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>Porcine epidemic diarrhea virus</td>\n",
       "      <td>species</td>\n",
       "      <td>9823</td>\n",
       "      <td>species</td>\n",
       "      <td>MTPLIYFWLFLPVLLTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...</td>\n",
       "      <td>NOT homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A8B1JDD3</td>\n",
       "      <td>2697049</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>QTZ33356.1</td>\n",
       "      <td>['Homo sapiens']</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>no rank</td>\n",
       "      <td>9606</td>\n",
       "      <td>species</td>\n",
       "      <td>SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A6G8HJ63</td>\n",
       "      <td>28295</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QIM40847.1</td>\n",
       "      <td>['pig']</td>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>Porcine epidemic diarrhea virus</td>\n",
       "      <td>species</td>\n",
       "      <td>9823</td>\n",
       "      <td>species</td>\n",
       "      <td>MKSLTYFWLFLPVLSTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...</td>\n",
       "      <td>NOT homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A0H4A793</td>\n",
       "      <td>28295</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AKN45969.1</td>\n",
       "      <td>['swine']</td>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>Porcine epidemic diarrhea virus</td>\n",
       "      <td>species</td>\n",
       "      <td>9823</td>\n",
       "      <td>species</td>\n",
       "      <td>MKSLTYFWLFLPVLSTLSLPQDVTRCSANTNFRRFFSKFNVQAPAV...</td>\n",
       "      <td>NOT homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id   tax_id uniprot_host_tax_ids embl_ref_id    embl_host_name virus_host_name                                       virus_name virus_taxon_rank  virus_host_tax_id virus_host_taxon_rank                                                seq human_binary_label\n",
       "0  A0A6G8HJ43    28295                  NaN  QIM40843.1           ['pig']      sus scrofa                  Porcine epidemic diarrhea virus          species               9823               species  MKSLTYFWLFLPVLSTLSLPQDVTRCSANTNFRRFFSKFNVQAPAV...   NOT homo sapiens\n",
       "1  A0A0G3FCN8    28295                  NaN  AKJ85723.1         ['swine']      sus scrofa                  Porcine epidemic diarrhea virus          species               9823               species  MTPLIYFWLFLPVLLTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...   NOT homo sapiens\n",
       "2  A0A8B1JDD3  2697049               [9606]  QTZ33356.1  ['Homo sapiens']    homo sapiens  Severe acute respiratory syndrome coronavirus 2          no rank               9606               species  SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...       homo sapiens\n",
       "3  A0A6G8HJ63    28295                  NaN  QIM40847.1           ['pig']      sus scrofa                  Porcine epidemic diarrhea virus          species               9823               species  MKSLTYFWLFLPVLSTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...   NOT homo sapiens\n",
       "4  A0A0H4A793    28295                  NaN  AKN45969.1         ['swine']      sus scrofa                  Porcine epidemic diarrhea virus          species               9823               species  MKSLTYFWLFLPVLSTLSLPQDVTRCSANTNFRRFFSKFNVQAPAV...   NOT homo sapiens"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/uniprot/coronaviridae_s_uniprot_uniprot_metadata_embl_hosts_pruned_metadata_corrected_species_virus_host_vertebrates_w_seq_t0.01_c8.csv\")\n",
    "df = pd.read_csv(input_file_path)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8fae834c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3568, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>seq</th>\n",
       "      <th>virus_name</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>virus_host_name</th>\n",
       "      <th>human_binary_label</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>uniprot_host_tax_ids</th>\n",
       "      <th>embl_ref_id</th>\n",
       "      <th>embl_host_name</th>\n",
       "      <th>virus_taxon_rank</th>\n",
       "      <th>virus_host_tax_id</th>\n",
       "      <th>virus_host_taxon_rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WIV04</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>WIV04(MN996528.1) Wuhan variant index virus</td>\n",
       "      <td>1273.0</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A0A6G8HJ43</td>\n",
       "      <td>MKSLTYFWLFLPVLSTLSLPQDVTRCSANTNFRRFFSKFNVQAPAV...</td>\n",
       "      <td>Porcine epidemic diarrhea virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>NOT homo sapiens</td>\n",
       "      <td>28295.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QIM40843.1</td>\n",
       "      <td>['pig']</td>\n",
       "      <td>species</td>\n",
       "      <td>9823.0</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A0A0G3FCN8</td>\n",
       "      <td>MTPLIYFWLFLPVLLTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...</td>\n",
       "      <td>Porcine epidemic diarrhea virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>NOT homo sapiens</td>\n",
       "      <td>28295.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>AKJ85723.1</td>\n",
       "      <td>['swine']</td>\n",
       "      <td>species</td>\n",
       "      <td>9823.0</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A8B1JDD3</td>\n",
       "      <td>SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...</td>\n",
       "      <td>Severe acute respiratory syndrome coronavirus 2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>2697049.0</td>\n",
       "      <td>[9606]</td>\n",
       "      <td>QTZ33356.1</td>\n",
       "      <td>['Homo sapiens']</td>\n",
       "      <td>no rank</td>\n",
       "      <td>9606.0</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A6G8HJ63</td>\n",
       "      <td>MKSLTYFWLFLPVLSTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...</td>\n",
       "      <td>Porcine epidemic diarrhea virus</td>\n",
       "      <td>NaN</td>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>NOT homo sapiens</td>\n",
       "      <td>28295.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>QIM40847.1</td>\n",
       "      <td>['pig']</td>\n",
       "      <td>species</td>\n",
       "      <td>9823.0</td>\n",
       "      <td>species</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id                                                seq                                       virus_name  seq_len virus_host_name human_binary_label     tax_id uniprot_host_tax_ids embl_ref_id    embl_host_name virus_taxon_rank  virus_host_tax_id virus_host_taxon_rank\n",
       "0       WIV04  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...      WIV04(MN996528.1) Wuhan variant index virus   1273.0    homo sapiens       homo sapiens        NaN                  NaN         NaN               NaN              NaN                NaN                   NaN\n",
       "0  A0A6G8HJ43  MKSLTYFWLFLPVLSTLSLPQDVTRCSANTNFRRFFSKFNVQAPAV...                  Porcine epidemic diarrhea virus      NaN      sus scrofa   NOT homo sapiens    28295.0                  NaN  QIM40843.1           ['pig']          species             9823.0               species\n",
       "1  A0A0G3FCN8  MTPLIYFWLFLPVLLTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...                  Porcine epidemic diarrhea virus      NaN      sus scrofa   NOT homo sapiens    28295.0                  NaN  AKJ85723.1         ['swine']          species             9823.0               species\n",
       "2  A0A8B1JDD3  SQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPF...  Severe acute respiratory syndrome coronavirus 2      NaN    homo sapiens       homo sapiens  2697049.0               [9606]  QTZ33356.1  ['Homo sapiens']          no rank             9606.0               species\n",
       "3  A0A6G8HJ63  MKSLTYFWLFLPVLSTLSLPQDVTRCQSTINFRRFFSKFNVQAPAV...                  Porcine epidemic diarrhea virus      NaN      sus scrofa   NOT homo sapiens    28295.0                  NaN  QIM40847.1           ['pig']          species             9823.0               species"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([wiv04_df, df])\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "263ab2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "unaligned_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/uniprot/aligned/coronaviridae_s_uniprot_uniprot_metadata_embl_hosts_pruned_metadata_corrected_species_virus_host_vertebrates_w_seq_t0.01_c8.csv\")\n",
    "df.to_csv(unaligned_file_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8074eef1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_aligned_dataset(unaligned_file_path, aligned_file_path):\n",
    "    print(unaligned_file_path)\n",
    "    unaligned_df = pd.read_csv(unaligned_file_path)[columns]\n",
    "    print(f\"unaligned_df size = {unaligned_df.shape}\")\n",
    "    # print(f\"Adding wiv04 to unaligned dataset\")\n",
    "    # unaligned_df = pd.concat([unaligned_df, wiv04_df])\n",
    "    # print(f\"unaligned_df size = {unaligned_df.shape}\")\n",
    "    aligned_df = parse_uniref_fasta_file(aligned_file_path)\n",
    "    print(f\"aligned_df size = {aligned_df.shape}\")\n",
    "    aligned_df.rename(columns={\"seq\": \"aligned_seq\"}, inplace=True)\n",
    "    df = aligned_df.merge(unaligned_df[columns], how=\"left\", on=\"uniprot_id\")\n",
    "    print(f\"merged df size = {df.shape}\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a6f8639",
   "metadata": {},
   "source": [
    "# Coronaviridae S Protein UniProt - EMBL mapping Aligned Dataset Generation\n",
    "## Coronaviridae S Protein"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2cdc45da",
   "metadata": {},
   "source": [
    "### coronaviridae_s_uniprot_embl_hosts_pruned_metadata_corrected_species_virus_host_vertebrates_w_seq_t0.01_c8_aligned.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "7abf2b3c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../../../../input/data/coronaviridae/20240313/uniprot/aligned/coronaviridae_s_uniprot_uniprot_metadata_embl_hosts_pruned_metadata_corrected_species_virus_host_vertebrates_w_seq_t0.01_c8.csv\n",
      "unaligned_df size = (3568, 4)\n",
      "START: Parsing fasta file\n",
      "/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../../../../input/data/coronaviridae/20240313/uniprot/aligned/coronaviridae_s_uniprot_uniprot_metadata_embl_hosts_pruned_metadata_corrected_species_virus_host_vertebrates_w_seq_t0.01_c8_aligned.fasta\n",
      "END: Parsing fasta file\n",
      "3568\n",
      "Number of records parsed = 3568\n",
      "aligned_df size = (3568, 2)\n",
      "merged df size = (3568, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniprot_id</th>\n",
       "      <th>aligned_seq</th>\n",
       "      <th>seq</th>\n",
       "      <th>virus_host_name</th>\n",
       "      <th>human_binary_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WIV04</td>\n",
       "      <td>-----MFVFLVLLPLVS---------------------------SQ...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P0DTC2</td>\n",
       "      <td>-----MFVFLVLLPLVS---------------------------SQ...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A0A679G9E9</td>\n",
       "      <td>-----MFVFLVLLPLVS---------------------------SQ...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A0A6M6IYS1</td>\n",
       "      <td>-----MFVFLVLLPLVS---------------------------SQ...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A0A7D5UC98</td>\n",
       "      <td>-----MFVFLVLLPLVS---------------------------SQ...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   uniprot_id                                        aligned_seq                                                seq virus_host_name human_binary_label\n",
       "0       WIV04  -----MFVFLVLLPLVS---------------------------SQ...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens\n",
       "1      P0DTC2  -----MFVFLVLLPLVS---------------------------SQ...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens\n",
       "2  A0A679G9E9  -----MFVFLVLLPLVS---------------------------SQ...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens\n",
       "3  A0A6M6IYS1  -----MFVFLVLLPLVS---------------------------SQ...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens\n",
       "4  A0A7D5UC98  -----MFVFLVLLPLVS---------------------------SQ...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligned_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/uniprot/aligned/coronaviridae_s_uniprot_uniprot_metadata_embl_hosts_pruned_metadata_corrected_species_virus_host_vertebrates_w_seq_t0.01_c8_aligned.fasta\")\n",
    "df = get_aligned_dataset(unaligned_file_path, aligned_file_path)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f21bfbd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/uniprot/aligned/coronaviridae_s_uniprot_uniprot_metadata_embl_hosts_pruned_metadata_corrected_species_virus_host_vertebrates_w_seq_t0.01_c8_aligned.csv\")\n",
    "df.to_csv(output_file_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "34add828",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique values = 8\n",
      "       virus_host_name  virus_host_name_count  virus_host_name_percent\n",
      "0           sus scrofa                   1626                45.571749\n",
      "1         homo sapiens                   1492                41.816143\n",
      "2          felis catus                    104                 2.914798\n",
      "3  camelus dromedarius                    103                 2.886771\n",
      "4           bos taurus                     83                 2.326233\n",
      "5        gallus gallus                     66                 1.849776\n",
      "6          canis lupus                     51                 1.429372\n",
      "7        bos grunniens                     43                 1.205157\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>virus_host_name</th>\n",
       "      <th>virus_host_name_count</th>\n",
       "      <th>virus_host_name_percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sus scrofa</td>\n",
       "      <td>1626</td>\n",
       "      <td>45.571749</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>1492</td>\n",
       "      <td>41.816143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>felis catus</td>\n",
       "      <td>104</td>\n",
       "      <td>2.914798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>camelus dromedarius</td>\n",
       "      <td>103</td>\n",
       "      <td>2.886771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bos taurus</td>\n",
       "      <td>83</td>\n",
       "      <td>2.326233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>gallus gallus</td>\n",
       "      <td>66</td>\n",
       "      <td>1.849776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>canis lupus</td>\n",
       "      <td>51</td>\n",
       "      <td>1.429372</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>bos grunniens</td>\n",
       "      <td>43</td>\n",
       "      <td>1.205157</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       virus_host_name  virus_host_name_count  virus_host_name_percent\n",
       "0           sus scrofa                   1626                45.571749\n",
       "1         homo sapiens                   1492                41.816143\n",
       "2          felis catus                    104                 2.914798\n",
       "3  camelus dromedarius                    103                 2.886771\n",
       "4           bos taurus                     83                 2.326233\n",
       "5        gallus gallus                     66                 1.849776\n",
       "6          canis lupus                     51                 1.429372\n",
       "7        bos grunniens                     43                 1.205157"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "column_stats(df, \"virus_host_name\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f87cb01",
   "metadata": {},
   "source": [
    "## Coronaviridae S protein - SARS-CoV-2 only\n",
    "### coronaviridae_s_uniref90_sars_cov_2_aligned.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "14fe610f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../../../../input/data/coronaviridae/20240313/uniref/sars_cov_2/coronaviridae_s_uniref90_sars_cov_2.csv\n",
      "unaligned_df size = (94, 4)\n",
      "START: Parsing fasta file\n",
      "/home/blessyantony/dev/git/zoonosis/src/jupyter_notebooks/datasets/generation/../../../../input/data/coronaviridae/20240313/uniref/sars_cov_2/coronaviridae_s_uniref90_sars_cov_2_aligned.fasta\n",
      "END: Parsing fasta file\n",
      "94\n",
      "Number of records parsed = 94\n",
      "aligned_df size = (94, 2)\n",
      "merged df size = (94, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>uniref90_id</th>\n",
       "      <th>aligned_seq</th>\n",
       "      <th>seq</th>\n",
       "      <th>virus_host_name</th>\n",
       "      <th>human_binary_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WIV04</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>UniRef90_A0A7U3HF24</td>\n",
       "      <td>------------------------PPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>PPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIHV...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>UniRef90_A0A7U3DWG5</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>UniRef90_A0A7U3EEK0</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>UniRef90_A0A7U3EEN6</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...</td>\n",
       "      <td>homo sapiens</td>\n",
       "      <td>homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           uniref90_id                                        aligned_seq                                                seq virus_host_name human_binary_label\n",
       "0                WIV04  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens\n",
       "1  UniRef90_A0A7U3HF24  ------------------------PPAYTNSFTRGVYYPDKVFRSS...  PPAYTNSFTRGVYYPDKVFRSSVLHSTQDLFLPFFSNVTWFHAIHV...    homo sapiens       homo sapiens\n",
       "2  UniRef90_A0A7U3DWG5  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens\n",
       "3  UniRef90_A0A7U3EEK0  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens\n",
       "4  UniRef90_A0A7U3EEN6  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...  MFVFLVLLPLVSSQCVNLTTRTQLPPAYTNSFTRGVYYPDKVFRSS...    homo sapiens       homo sapiens"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligned_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/uniprot/sars_cov_2/coronaviridae_s_uniprot_sars_cov_2_aligned.fasta\")\n",
    "unaligned_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/uniprot/sars_cov_2/coronaviridae_s_uniprot_sars_cov_2.csv\")\n",
    "df = get_aligned_dataset(unaligned_file_path, aligned_file_path)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7f461028",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[\"seq_len\"] = df[\"aligned_seq\"].apply(lambda x: len(str(x)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4580d843",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    1273\n",
       "1    1273\n",
       "2    1273\n",
       "3    1273\n",
       "4    1273\n",
       "Name: seq_len, dtype: int64"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[\"seq_len\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "17048179",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligned_csv_file_path = os.path.join(os.getcwd(), \"..\", \"..\", \"..\",\"..\", \"input/data/coronaviridae/20240313/uniref/sars_cov_2/coronaviridae_s_uniref90_sars_cov_2_aligned.csv\")\n",
    "df.to_csv(aligned_csv_file_path, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77d3b6f6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
